message.go

```diff
@@ -17,129 +14,40 @@ type Message struct {
+	AttachURL         string `json:"attachurl,omitempty"` // URL of an attachment
 }
 
 func (m *Message) MarshalJSON() ([]byte, error) {
-	buf, err := json.Marshal(m.Topic)
-	if err != nil {
-		return nil, err
-	}
-	buf = []byte(fmt.Sprintf(`{"topic":%s`, buf))
-
-	if x := m.Message; x != "" {
-		mm, err := json.Marshal(x)
-		if err != nil {
-			return nil, err
-		}
-		buf = append(buf, fmt.Sprintf(`,"message":%s`, mm)...)
-	}
-
-	if x := m.Title; x != "" {
-		mm, err := json.Marshal(x)
-		if err != nil {
-			return nil, err
-		}
-		buf = append(buf, fmt.Sprintf(`,"title":%s`, mm)...)
-	}
-
-	if x := m.Tags; len(x) > 0 {
-		mm, err := json.Marshal(x)
-		if err != nil {
-			return nil, err
-		}
-		buf = append(buf, fmt.Sprintf(`,"tags":%s`, mm)...)
-	}
-
-	if x := m.Priority; x > 0 {
-		mm, err := json.Marshal(x)
-		if err != nil {
-			return nil, err
-		}
-		buf = append(buf, fmt.Sprintf(`,"priority":%s`, mm)...)
-	}
-
-	if x := m.Actions; len(x) > 0 {
-		mm, err := json.Marshal(x)
-		if err != nil {
-			return nil, err
-		}
-		buf = append(buf, fmt.Sprintf(`,"actions":%s`, mm)...)
-	}
-
-	type urls struct {
-		name string
-		url  *url.URL
-	}
-
-	for _, v := range []urls{
-		{"click", m.ClickURL},
-		{"attachurl", m.AttachURL},
-		{"icon", m.IconURL},
-	} {
-		mm, err := urlString(v.url)
-		if err != nil {
-			return nil, err
-		}
-
-		if mm == nil {
-			continue
-		}
-
-		buf = append(buf, fmt.Sprintf(`,"%s":%s`, v.name, mm)...)
-	}
-
-	if x := m.Delay; x > 0 {
-		mm, err := json.Marshal(x.String())
-		if err != nil {
-			return nil, err
-		}
-		buf = append(buf, fmt.Sprintf(`,"delay":%s`, mm)...)
-	}
-
-	if x := m.Email; x != `` {
-		mm, err := json.Marshal(x)
-		if err != nil {
-			return nil, err
-		}
-		buf = append(buf, fmt.Sprintf(`,"email":%s`, mm)...)
-	}
-
-	if x := m.Call; x != `` {
-		mm, err := json.Marshal(x)
-		if err != nil {
-			return nil, err
-		}
-		buf = append(buf, fmt.Sprintf(`,"call":%s`, mm)...)
-	}
-
-	if x := m.AttachURLFilename; x != `` {
-		mm, err := json.Marshal(x)
-		if err != nil {
-			return nil, err
-		}
-		buf = append(buf, fmt.Sprintf(`,"filename":%s`, mm)...)
-	}
-
-	return append(buf, '}'), nil
-}
-
-func urlString(u *url.URL) ([]byte, error) {
-	if u == nil {
-		return nil, nil
-	}
-
-	s := u.String()
-	if s == "" {
-		return nil, nil
-	}
-
-	return json.Marshal(s)
+	// if m.Priority <0, drop it
+	if m.Priority < 0 {
+		m.Priority = 0 // can't use nil, but this works
+	}
+	// if m.Delay is int and <=0, drop it
+	if i, ok := m.Delay.(int); ok && i <= 0 {
+		m.Delay = nil
+	}
+
+	return json.Marshal(Message{
+		Topic:    m.Topic,
+		Message:  m.Message,
+		Title:    m.Title,
+		Tags:     m.Tags,
+		Priority: m.Priority,
+		Actions:  m.Actions,
+		ClickURL: m.ClickURL,
+		IconURL:  m.IconURL,
+		Delay:    m.Delay,
+		Email:    m.Email,
+		Call:     m.Call,
+
+		AttachURLFilename: m.AttachURLFilename,
+		AttachURL:         m.AttachURL,
+	})
```

Comment:
I can't recall if this would invoke recursion. I don't think it does because you're passing a struct not a pointer. If it does invoke recursion however, we obviously have a problem because this is an infinite loop so this may require `map[string]any`

Reply:
So this is fine? it works fine for me in tests and in a project (added `replace github.com/AnthonyHewins/gotfy => github.com/SConaway/gotfy v0.1.11` to `go.mod`)

Reply:
Yeah I just checked to be sure